lib/AST/ProtocolGraph.cpp

```diff
@@ -0,0 +1,214 @@
+  // Visit protocols in reverse order, so that if P inherits from Q and
+  // Q inherits from R, we first visit R, then Q, then P, ensuring that
+  // R's associated types are added to P's list, etc.
+  for (const auto *proto : llvm::reverse(Protocols)) {
+    auto &info = Info[proto];
+
+    // We might inherit the same associated type multiple times due to
+    // diamond inheritance, so make sure we only add each associated
+    // type once.
+    llvm::SmallDenseSet<const AssociatedTypeDecl *, 4> visited;
+
+    for (const auto *inherited : info.Inherited) {
+      if (inherited == proto)
+        continue;
+
+      for (auto *inheritedType : getProtocolInfo(inherited).AssociatedTypes) {
+        if (!visited.insert(inheritedType).second)
+          continue;
+
+        // The 'if (inherited == proto)' above avoids a potential
+        // iterator invalidation here, because we're updating
+        // getProtocolInfo(proto).AssociatedTypes while iterating over
+        // getProtocolInfo(inherited).AssociatedTypes.
+        info.AssociatedTypes.push_back(inheritedType);
+      }
+    }
+  }
+}
+
+// Update each protocol's AllInherited vector to add all transitively
+// inherited protocols.
+void ProtocolGraph::computeInheritedProtocols() {
+  // Visit protocols in reverse order, so that if P inherits from Q and
+  // Q inherits from R, we first visit R, then Q, then P, ensuring that
+  // R's associated types are added to P's list, etc.
+  for (const auto *proto : llvm::reverse(Protocols)) {
+    auto &info = Info[proto];
+
+    // We might inherit the same protocol multiple times due to diamond
+    // inheritance, so make sure we only add each protocol once.
+    llvm::SmallDenseSet<const ProtocolDecl *, 4> visited;
+    visited.insert(proto);
+
+    for (const auto *inherited : info.Inherited) {
+      // Add directly-inherited protocols.
+      if (!visited.insert(inherited).second)
+        continue;
+      info.AllInherited.push_back(inherited);
+
+      // Add indirectly-inherited protocols.
+      for (auto *inheritedType : getProtocolInfo(inherited).AllInherited) {
+        if (!visited.insert(inheritedType).second)
+          continue;
+
+        info.AllInherited.push_back(inheritedType);
+      }
+    }
+  }
+}
+
+/// Recursively compute the 'depth' of e protocol, which is inductively defined
+/// as one greater than the depth of all inherited protocols, with a protocol
+/// that does not inherit any other protocol having a depth of one.
+unsigned ProtocolGraph::computeProtocolDepth(const ProtocolDecl *proto) {
+  auto &info = Info[proto];
+
+  if (info.Mark) {
+    // Already computed, or we have a cycle. Cycles are diagnosed
+    // elsewhere in the type checker, so we don't have to do
+    // anything here.
+    return info.Depth;
+  }
+
+  info.Mark = true;
+  unsigned depth = 0;
+
+  for (auto *inherited : info.Inherited) {
+    unsigned inheritedDepth = computeProtocolDepth(inherited);
```

Comment:
Refactoring this to use a worklist is going to be a little bit tricky. We recurse over inherited protocols in other places too. Note that we should already be able to detect cycles here, because of the 'Mark' bit.